src/utils/io.py

```diff
@@ -1,34 +1,229 @@
-"""Module handling file input/output"""
+"""Database I/O utilities for CSV and Parquet formats."""
```

Comment:
I did not review the IO stuff all that carefully, because it's hard for me to follow and it looks like you tested it carefully. I'm also assuming that you considered different high-level approaches and chunking like this was the best option. If it works, I'm happy!